modules/rubiconBidAdapter.js

```diff
@@ -503,51 +492,45 @@ export const spec = {
-      const liveintentId = find(bidRequest.userIdAsEids, eid => eid.source === 'liveintent.com');
-      if (liveintentId) {
-        data['tpid_liveintent.com'] = liveintentId.uids[0].id;
-        if (liveintentId.ext && Array.isArray(liveintentId.ext.segments) && liveintentId.ext.segments.length) {
-          data['tg_v.LIseg'] = liveintentId.ext.segments.join(',');
-        }
-      }
-      const liverampId = find(bidRequest.userIdAsEids, eid => eid.source === 'liveramp.com');
-      if (liverampId) {
-        data['x_liverampidl'] = liverampId.uids[0].id;
-      }
-      const sharedId = find(bidRequest.userIdAsEids, eid => eid.source === 'sharedid.org');
-      if (sharedId) {
-        data['eid_sharedid.org'] = `${sharedId.uids[0].id}^${sharedId.uids[0].atype}^${sharedId.uids[0].ext.third}`;
-      }
-      const pubcid = find(bidRequest.userIdAsEids, eid => eid.source === 'pubcid.org');
-      if (pubcid) {
-        data['eid_pubcid.org'] = `${pubcid.uids[0].id}^${pubcid.uids[0].atype}`;
-      }
-      const criteoId = find(bidRequest.userIdAsEids, eid => eid.source === 'criteo.com');
-      if (criteoId) {
-        data['eid_criteo.com'] = `${criteoId.uids[0].id}^${criteoId.uids[0].atype}`;
-      }
-    }
-
-    // set ppuid value from config value
+    // pass publisher provided userId if configured
     const configUserId = config.getConfig('user.id');
     if (configUserId) {
       data['ppuid'] = configUserId;
-    } else {
-      // if config.getConfig('user.id') doesn't return anything, then look for the first eid.uids[*].ext.stype === 'ppuid'
-      for (let i = 0; bidRequest.userIdAsEids && i < bidRequest.userIdAsEids.length; i++) {
-        if (bidRequest.userIdAsEids[i].uids) {
-          const pubProvidedId = find(bidRequest.userIdAsEids[i].uids, uid => uid.ext && uid.ext.stype === 'ppuid');
-          if (pubProvidedId && pubProvidedId.id) {
-            data['ppuid'] = pubProvidedId.id;
-            break;
+    }
+    // loop through userIds and add to request
+    if (bidRequest.userIdAsEids) {
+      bidRequest.userIdAsEids.forEach(eid => {
+        try {
+          // special cases
+          if (eid.source === 'adserver.org') {
+            data['tpid_tdid'] = eid.uids[0].id;
+            data['eid_adserver.org'] = eid.uids[0].id;
+          } else if (eid.source === 'liveintent.com') {
+            data['tpid_liveintent.com'] = eid.uids[0].id;
+            data['eid_liveintent.com'] = eid.uids[0].id;
+            if (eid.ext && Array.isArray(eid.ext.segments) && eid.ext.segments.length) {
+              data['tg_v.LIseg'] = eid.ext.segments.join(',');
+            }
+          } else if (eid.source === 'liveramp.com') {
+            data['x_liverampidl'] = eid.uids[0].id;
+          } else if (eid.source === 'sharedid.org') {
+            data['eid_sharedid.org'] = `${eid.uids[0].id}^${eid.uids[0].atype}^${eid.uids[0].ext.third}`;
+          } else {
+            // add anything else with this generic format
```

Comment:
just a note that ID5 module includes an `ext` that contains additional metadata about the ID (see https://github.com/prebid/Prebid.js/blob/master/modules/userId/eids.md). Currently, we are including `linkType` which signals the method that we used to link an ID across domains, something that buyers have asked to be included in bid requests to them, but we may add more metadata later that should be passed through.

@jdwieland8282 this is roughly equivalent to the `stype` conversation happening related to pub provided ids, so feels important to be included here.

Reply:
Thanks @smenzer.  Need to see if our endpoint supports this.  If not, may do this in another PR after we support it.

Reply:
thanks @smenzer - will see if our endpoint supports this.  if not, will make the change in a separate PR after we support it.

Reply:
made the update @smenzer

Reply:
looks good, thanks!